src/connection-loop.ts

```diff
@@ -0,0 +1,261 @@
+  //
+  // The loop flows through the following states forever, until it is closed:
+  //
+  // Pending: Wait for event or watchdog
+  //          |
+  //          v
+  // Debounce: Wait for more events (we debounce pushes)
+  //          |
+  //          v
+  // Wait for available connection (we limit number of parallel requests
+  // allowed)
+  //          |
+  //          v
+  // Wait to send (if requests are taking too long, we will slow down)
+  //          |
+  //          v
+  // Send (asynchronously, wrt the loop)
+  //          |
+  //          v
+  // Back to the pending!
+
+  // Controls whether the next iteration of the loop will wait at the pending
+  // state.
+  private _pendingResolver = resolver<void>();
+
+  // Controls when send() will return. This is purely a convenience to callers
+  // who might want to know when the related push/pull operation actually
+  // competed without using the onSync API.
+  private _currentResolver = resolver<void>();
+
+  private readonly _delegate: ConnectionLoopDelegate;
+  private _closed = false;
+
+  constructor(delegate: ConnectionLoopDelegate) {
+    this._delegate = delegate;
+    this.run();
+  }
+
+  close(): void {
+    this._closed = true;
+  }
+
+  async send(): Promise<void> {
+    this._pendingResolver.resolve();
+    await this._currentResolver.promise;
+  }
+
+  async run(): Promise<void> {
+    const sendRecords: SendRecord[] = [];
+    let delay = MIN_DELAY;
+    let recoverResolver = resolver();
+    let lastSendTime = 0;
+
+    // The number of active connections.
+    let counter = 0;
+    const delegate = this._delegate;
+    const {
+      debounceDelay = () => DEBOUNCE_DELAY,
+      maxConnections = () => MAX_CONNECTIONS,
+      watchdogTimer,
+      log,
+    } = delegate;
+
+    log?.('Starting connection loop');
+
+    while (!this._closed) {
+      log?.(
+        didLastSendRequestFail(sendRecords)
+          ? 'Last request failed. Trying again'
+          : 'Waiting for a send',
+      );
+
+      // The current resolver is used to make the individual calls to exeute
+      // have the correct "duration".
+      const currentResolver = resolver();
+      this._currentResolver = currentResolver;
+      // Don't let this rejected promise escape.
+      currentResolver.promise.catch(() => 0);
+
+      // Wait until send is called or until the watchdog timer fires.
+      const races = [this._pendingResolver.promise];
+      if (watchdogTimer) {
+        races.push(sleep(watchdogTimer()));
+      }
+      await Promise.race(races);
+
+      await sleep(debounceDelay());
+
+      // This resolver is used to wait for incoming push calls.
+      this._pendingResolver = resolver();
+
+      if (counter >= maxConnections()) {
+        log?.('Too many pushes. Waiting until one finishes...');
+        await this._waitUntilAvailableConnection();
+        log?.('...finished');
+      }
+
+      // We need to delay the next request even if there are no active requests
```

Comment:
Can you help me understand the problems this is solving (maybe in a comment on computeDelayAndUpdateDurations)? It looks like the intent is:
- exponentially back off if the server is erroring so as not to hammer it.  And when the server stops erroring, go back to a normal delay.
- ensure consist rate of connections by spacing requests out according to the median request time over the last N requests (ie, if median time is t send a request every t / # connections).